## Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

### Step 1. Open the data file and have a look at the general information. 

In [1]:
import pandas as pd
from nltk.stem import SnowballStemmer
english_stemmer = SnowballStemmer('english')
bank_data = pd.read_csv('/datasets/credit_scoring_eng.csv') #to read the file into dataframe
bank_data.head(10) #to check first 10  columns of data

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

In [2]:
bank_data.info() #to get the entire information of the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
bank_data.isna().sum() #to detect missing & undefined values in the dataframe

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

### Conclusion

The given dataset surely comprises of many columns than that we worked out in th course tasks which is a good thing, as it is a bit connected to real life dataset which is to be analysed in coming days. The first ten rows of the data have been printed out and the data in it looks pretty good so far except for the days_employed column due to it's negative values. That can be analysed further.

There are 12 columns in the data out of which :

1. 'children','days_employed','dob_years','total_income' columns represent the numerical values so thier datatype should be int64 for an easy analysis.

 However, from the information given in dataframe the columns 'days_employed' and 'total_income' have float64 datatype.

2. 'education','family_status','gender','income_type','purpose' columns give us information in words so thier datatype should be object.

      
3. 'eduation_id','family_status_id','debt' columns give us information about the presence/absence of IDs and debts in '0' & '1' form, considering 0-present , 1-absent. So their datatype should be 'int'


4. As expected the real life data always has some impurities in it like missing values, duplicates, typo errors..etc. In our data we have missing value in two columns i.e., 'days_employed' and 'total_income' which are equal to 2174 each.
 


### Step 2. Data preprocessing

### Processing missing values

In [4]:
bank_data.isna().sum() #to detect missing & undefined values in the dataframe

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

----- It seems like the 'days_employed' and 'total_income' columns have 2174 missing values each. So let's analyse each column and get some more info. ------

In [5]:
bank_data['days_employed'].value_counts()

-986.927316     1
-7026.359174    1
-4236.274243    1
-6620.396473    1
-1238.560080    1
               ..
-2849.351119    1
-5619.328204    1
-448.829898     1
-1687.038672    1
-582.538413     1
Name: days_employed, Length: 19351, dtype: int64

In [6]:
bank_data['days_employed'].head(30)

0      -8437.673028
1      -4024.803754
2      -5623.422610
3      -4124.747207
4     340266.072047
5       -926.185831
6      -2879.202052
7       -152.779569
8      -6929.865299
9      -2188.756445
10     -4171.483647
11      -792.701887
12              NaN
13     -1846.641941
14     -1844.956182
15      -972.364419
16     -1719.934226
17     -2369.999720
18    400281.136913
19    -10038.818549
20     -1311.604166
21      -253.685166
22     -1766.644138
23      -272.981385
24    338551.952911
25    363548.489348
26              NaN
27      -529.191635
28      -717.274324
29              NaN
Name: days_employed, dtype: float64

In [7]:
bank_data['days_employed'].tail(30)

21495              NaN
21496      -759.680851
21497              NaN
21498     -1330.627998
21499     -9929.015065
21500      -578.082757
21501    334343.096304
21502              NaN
21503     -3096.881131
21504    355235.728158
21505    338904.866406
21506     -1556.249906
21507       -79.832064
21508    386497.714078
21509    362161.054124
21510              NaN
21511      -612.569129
21512      -165.377752
21513     -1166.216789
21514      -280.469996
21515      -467.685130
21516      -914.391429
21517      -404.679034
21518    373995.710838
21519     -2351.431934
21520     -4529.316663
21521    343937.404131
21522     -2113.346888
21523     -3112.481705
21524     -1984.507589
Name: days_employed, dtype: float64

As 'days_employed' column had some missing values , I have used .value_counts() to check what are the exact values present in this column and when I checked the first and last 30 columns of the data, I noticed that maximum values in it are negative. This surely shows that further analysis is required.

In [8]:
bank_data['days_employed'].max()

401755.40047533

In [9]:
bank_data['days_employed'].min()

-18388.949900568383

------Let's check for the maximum and minimum values of the data so that we can get an idea on how exactly the data makes sense?------

If we try to convert the maximun value i.e., 401755.4 to years considering it to be days then the maximum years of days_employed would be 1,100.69 years which is like crazy ;D 

on the other hand when we convert -18388.94 to years and consider it to be over estimated even then the data doesn't make any sense because that would make over 600 years which would be absolutely meaningless !!!

I would like to leave the column as it is because it isn't much relevant to what we are currently looking for.

In [10]:
bank_data['total_income'].value_counts() # to display unique values and their counts

17312.717    2
31791.384    2
42413.096    2
54857.666    1
26935.722    1
            ..
48796.341    1
34774.610    1
15710.698    1
19232.334    1
9591.824     1
Name: total_income, Length: 19348, dtype: int64

--------The 'total_income' column seems to be pretty reasonable and the data appears to be genuine as everything is in normal income range of thousands yet there are 2174 missing values in it.--------

In [11]:
bank_data['total_income'].max()

362496.645

In [12]:
bank_data['total_income'].min()

3306.762

In [13]:
income_median = bank_data['total_income'].median()

In [14]:
bank_data['total_income'] = bank_data['total_income'].fillna(income_median) # to fill the missing values with 0 in 'total_income'
bank_data['total_income'].isna().sum()

0

-----------I decided to fill in the missing values in 'total_income' column with it's median because when we look at the maximum and minimum values of the column the difference is pretty big. It can't be filled with '0' unless all the users are not employed and neither with mean value because there might be few outliers in the column as the difference between the max and min values are huge . So I decided to replace it with median which would give me the accurate data for analysis.----------

In [15]:
bank_data['children'].value_counts() # to display unique values and their counts

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [16]:
bank_data.loc[(bank_data['children']==-1),'children'] = 1 # to replace all -1 to 1 in the 'children' column
bank_data.loc[(bank_data['children']==20),'children'] = 2 # to replace all 20 to 2 in the 'children' column
bank_data['children'].value_counts() # to display unique values and their counts

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

------Coming to the 'children' column it's strange to find a -1 and 20 in it. Probably -1 may be result of some kind of typing error so I would like to replace it by it's absolute value 1. Now considering the value 20 , in general people these days are'nt opting for so many kids even if there are few exceptional cases surely their number isn't gonna be as big as 76. So I would like to replace it by 2 considering 0 to be a typo.--------

In [17]:
bank_data['family_status'].value_counts() # to display unique values and their counts

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

-------------The 'family_status' column seems to be pretty much perfect.. so it remains absolutely UNTOUCHED :D   ----------

In [18]:
bank_data['purpose'].value_counts() # to display unique values and their counts

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
housing transactions                        653
buying property for renting out             653
transactions with commercial real estate    651
purchase of the house                       647
housing                                     647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
building a property                         620
purchase of my own house                    620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

-------------The 'purpose' column seems to be pretty much perfect.. so it remains absolutely UNTOUCHED :D  ----------

In [19]:
bank_data['dob_years'].value_counts() # to display unique values and their counts

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

---------------  I feel like this column can be pretty decieving if it isn't properly looked at. Almost at the bottom part of value_counts() output we can see that 101 times the age 0 is repeated. How can someone earn something or do something without even being born .. ;P   ---------------


In [20]:
replace_by = bank_data['dob_years'].median() #calculating median of the 'dob_years' column
(bank_data.loc[bank_data['dob_years']==0 ,'dob_years'])= replace_by #to fill in '0' with the median value
bank_data['dob_years'].value_counts() # to display unique values and their counts


42.0    698
35.0    617
40.0    609
41.0    607
34.0    603
38.0    598
33.0    581
39.0    573
31.0    560
36.0    555
44.0    547
29.0    545
30.0    540
48.0    538
37.0    537
50.0    514
43.0    513
32.0    510
49.0    508
28.0    503
45.0    497
27.0    493
56.0    487
52.0    484
47.0    480
54.0    479
46.0    475
58.0    461
57.0    460
53.0    459
51.0    448
59.0    444
55.0    443
26.0    408
60.0    377
25.0    357
61.0    355
62.0    352
63.0    269
64.0    265
24.0    264
23.0    254
65.0    194
66.0    183
22.0    183
67.0    167
21.0    111
68.0     99
69.0     85
70.0     65
71.0     58
20.0     51
72.0     33
19.0     14
73.0      8
74.0      6
75.0      1
Name: dob_years, dtype: int64

----------------   So, I decided to replace 0 with the median value rather than mean value because there is a huge difference between the maximum and minimum age being 75 and 19 respectively.    --------------

In [21]:
bank_data['debt'].value_counts() # to display unique values and their counts

0    19784
1     1741
Name: debt, dtype: int64

  -------------------      Seems like maximum people are prompt at clearing their past debts :)    ----------------

In [22]:
bank_data['education'].value_counts() # to display unique values and their counts

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
GRADUATE DEGREE            1
Graduate Degree            1
Name: education, dtype: int64

-----  This column may need some case modifications like converting to lower case rather than upper case because the other columns of the data are in lower case.    -------

In [23]:
bank_data['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

------ I don't really have any idea of what XNA exactly is !! as the number of count is just 1 out of 21,525 values , I would prefer to ignore it..! ;)  ------

### Conclusion

1. There were few missing values in the columns 'days_employed' and 'total_income' which were surprisingly equal to 2174

2. The 'children' column had no missing values yet had some unusual values like -1 , 20 which in  my view would be typo errors because having -1 child is pointless so I assumed it to be absolute value 1 and hence replaced it and if we  look in a broad aspect these days having 20 kids is quite rare so i replaced it with 2 children.

3. Coming to the 'days_employed' and 'total_income' columns:

  a. in'total_income' column I decided to fill up all the missing values with median .
  
  b. 'days_employed' column seems to be highly corrupted because no matter what way I use trying to convert to years or something else the data doesn't make sense.
  
4. In the 'dob_years' column 101 members age is 0.. which can't be real in any aspect, I would prefer replacing it with median value rather than dropping it assuming it to be a typo error.
  
5. Apart from these I think the data is free from missing values and I would like to talk to the data engineers or tech support about the strange values in the 'days_employed' , 'total_income' , 'gender' and 'age' columns and find a way to prevent getting such anomalies in the future.

### Data type replacement

In [24]:
bank_data.info() #to get the entire information of the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null float64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null float64
purpose             21525 non-null object
dtypes: float64(3), int64(4), object(5)
memory usage: 2.0+ MB


----->  From the output obtained above: 

 1. all the quantitative variables namely 'children' , 'education_id' , 'family_status_id' , 'debt' are in their desired datatype , except for 'days_employed' , 'dob_years' and 'total_income' are needed to be converted to it's desired data type 'int'.
 
 
 2. all the categorical variables namely 'education' , 'family_status' , 'gender' , 'income_type' and 'purpose' are in their desired data type i.e., object.

In [25]:
bank_data['total_income'] = bank_data['total_income'].astype('int') #to change the datatype to integer of 'total_income' column
bank_data['total_income'].head(10)#to make sure the datatype is changed

0        40620
1        17932
2        23341
3        42820
4        25378
         ...  
21520    35966
21521    24959
21522    14347
21523    39054
21524    13127
Name: total_income, Length: 21525, dtype: int64

In [26]:
bank_data['dob_years'] = bank_data['dob_years'].astype('int')
bank_data['dob_years'].head(10)

0        42
1        36
2        33
3        32
4        53
         ..
21520    43
21521    67
21522    38
21523    38
21524    40
Name: dob_years, Length: 21525, dtype: int64

### Conclusion

1. The data types of different columns in the dataframe seem to be quite ready to analyse except for the 'days_employed' , 'dob_years' and 'total_income' columns.


2. The 'days_employed' column remains untouched as the data provided seems to be corrupted and the datatype of 'total_income' and  'dob_years' columns is 'float64' which is needed to be converted to 'int'.


3. In this case, when we change the data type of 'total_income' from float to int, the values are rounded off which doesn't show much of the impact on our analysis.

### Processing duplicates

In [27]:
len(bank_data[bank_data.duplicated()]) #checking for total number of duplicates in the dataframe

55

--------  Initially, .duplicated() method gave us the information that there are 55 duplicates in the data. But let us analyse our data further..!!  --------

In [28]:
bank_data['education'].value_counts() # to display unique values and their counts

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
GRADUATE DEGREE            1
Graduate Degree            1
Name: education, dtype: int64

--------------From the above output we can observe that there are similar types of education printed in upper and lower cases randomly, this may be because of multiple people entering different data into the database..! We need to convert all the alphabets of 'education' column to lower case because the enitre data is in lower case.------------

In [29]:
bank_data['education'] = bank_data['education'].str.lower() #converting all elements in the 'education' column to lowercase to ease the duplicating process.

In [30]:
len(bank_data[bank_data.duplicated()]) #checking for total number of duplicates in the dataframe

72

----------- Ahaa..!!! The total number of duplicates have been increased from 55 to 72 after changine the alphabets case. -------------

In [31]:
bank_data.drop_duplicates().head(10) #dropping off the duplicate rows by just keeping the original one.

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  secondary education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0             40620    purchase of the house  
1             17932             car purchase  
2             23341    purchase of the house  
3             42820  supplementary education  
4             25378        to have a wedding  
...             ...                      ...  
21520         35966     housing transactions  
21521         24959        purchase of a car  
21522         14347                 property  
21523         39054        buying my own car  
21524         13127             to buy a car  

[21453 rows x 12 columns]

### Conclusion

The dataframe had 55 duplicate values, some of these duplicates might be detected due to multiple entries of the same data repeatedly , so I just took a broad look at the data and found out that converting all alphabets in the 'education' column to lower case resulted in a much accurate data , there by increasing the duplicates to 72.

Then applying the .drop_duplicates() method and voila! the rows have reduced to 21453. Now our data is free from dupicates.

The reason for appearence of these duplicates in the data might be due to multiple people involved in entering the data to the system or sometimes just to bluff the company with multiple entries. Such detections must be reported to the data engineers/tech support group in the company so that such mistakes don't occur repeatedly. 

### Categorizing Data

In [32]:
bank_data.info() #to get the general information of the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


Ok..!! Now our data is ready to be processed further as there are no missing values, duplicates are eliminated and the columns have their data types assigned correctly..!!

In [33]:
children_pivot = bank_data.pivot_table(index='children',values='debt',aggfunc='mean')
children_pivot

debt
children          
0         0.075129
1         0.091470
2         0.094791
3         0.081818
4         0.097561
5         0.000000

In [34]:
children_with_debts = children_pivot['debt']*100
children_with_debts

children
0    7.512898
1    9.146968
2    9.479118
3    8.181818
4    9.756098
5    0.000000
Name: debt, dtype: float64

Now, we need to calculate the impact of number of children in a family for the on-time repayment of loan therefore we need to check the precentage of families with different number of children ranging from 0 to 5 still possess debts that are yet to be cleared. So, i created a pivot_table() which takes the index as 'children' , the values as'debt' and aggregate function as mean because the column debts doesn't have any outliers.


Then we calculate the percentages of the mean values of debts corresponding to the number of children in the family. I decided to calculate the percentages because I feel it's good to get things to be compared to a neutral form and then comparing it.

In [35]:
family_status_pivot = bank_data.pivot_table(index='family_status',values='debt',aggfunc= 'mean')
family_status_pivot

debt
family_status              
civil partnership  0.092890
divorced           0.071130
married            0.075202
unmarried          0.097405
widow / widower    0.065625

In [36]:
family_status_with_debt = family_status_pivot['debt']*100
family_status_with_debt 

family_status
civil partnership    9.288963
divorced             7.112971
married              7.520194
unmarried            9.740491
widow / widower      6.562500
Name: debt, dtype: float64

Coming to the 'family_status' column , I analysed it just as I did above in the analysis  of impact of children. I just created a pivot_table() showing the family_status column's unique values and their corresponding debts mean value. Then I converted them into percentages for better comparision.

In [37]:
#categorizing data based on total_income and debt:
#creating a new function 

def total_income(row):
    income = row['total_income']
    
    if income <= 21682:
        return 'less income'
    else:
        return 'high income'

The 'total_income' column has several unique values. So, I think it's better we divide them into two categories of high and low incomes. Let's formulate a function, by putting the comparing parameter to be the median value of the column i.e., 21682 families earning above this income we put into high income and others into less income category.

In [38]:
bank_data['total_income_category'] = bank_data.apply(total_income,axis=1) #applying the function to dataframe 
bank_data.head(10)

children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  secondary education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose total_income_category  
0             40620    purchase of the house           high income  
1             17932             car purchase           less income  
2             23341    purchase of the house           high income  
3             42820  supplementary education           high income  
4             25378        to have a wedding           high income  
...             ...                      ...                   ...  
21520         35966     housing transactions           high income  
21521         24959        purchase of a car           high income  
21522         14347                 property           less income  
21523         39054        buying my own car           high income  
21524         13127             to buy a car           less income  

[21525 rows x 13 columns]

Let's check whether our "total_income" function works or not. 

Apply the function to the data frame using .apply() and putting axis=1 so that each row is examined one by one and store the resultant values into a new column namely 'total_income_category'.

In [39]:
income_pivot = bank_data.pivot_table(index='total_income_category',values='debt',aggfunc='mean')
income_pivot

debt
total_income_category          
high income            0.080550
less income            0.081383

In [40]:
income_with_debt = income_pivot['debt']*100
income_with_debt

total_income_category
high income    8.055040
less income    8.138316
Name: debt, dtype: float64

Now we need to construct a pivot_table() showing the 'total_income_category' column's unique values and their corresponding debts mean value. Then I converted them into percentages for better comparision. 

In [41]:
#categorizing data based on the stemm words in 'purpose' 
##creating a new function for stemming process

def getStem(row):
    stemm_word = ['wed','hous','estat','properti','construct','car','educ']
    for word in row['purpose'].split(" "):
        stemmed_word = english_stemmer.stem(word)
        for value in stemm_word:
            if stemmed_word == 'hous' or stemmed_word =='estat' or stemmed_word=='properti':
                row['purpose_stem'] = 'construct'
                return row
            if stemmed_word == value:
                row['purpose_stem'] = value
                return row
           

In the 'purpose' column there are several unique values which can be categorised into one, just like i did above by converting 'hous','estat','properti' and 'construct' in to a single category 'construct' to simplify my analysis and also the other purposes matching the stemm words like 'wed','car','educ',then stocking them up into a new column named as 'purpose_stem'.

In [42]:
#categoriyzing data based on stemmed words and debts
#creating a new function

def purpose(row):
    status = row['purpose_stem']
    no_debts = row['debt']
    stemm_word = ['wed','hous','car','educ']
    if no_debts == 0:
        for value in stemm_word:
            if status == value:
                return value +' and debt free'
    if no_debts == 1:
        for value in stemm_word:
            if status == value:
                return value +' with debts'

The function 'purpose' has been defined to compare the unique values in the'purpose_stem' column to the values in the 'debt' column.

### Conclusion

The categorization of data has been done on the basis of aspects that were asked to be analyzed by the bank like 'children , 'family_status' , 'total_income' and 'purpose' that affect the repayment of loan thereby using it during the process of preparing the credit score.

### Step 3. Answer these questions

- Is there a relation between having kids and repaying a loan on time?

In [43]:
children_with_debts = children_pivot['debt']*100
children_with_debts

children
0    7.512898
1    9.146968
2    9.479118
3    8.181818
4    9.756098
5    0.000000
Name: debt, dtype: float64

### Conclusion

As per the above results, it's evident that number of children present in the family definitely has an impact on repayment of debts. This is proved by the above results, as we can see the highest percentage of debt holders is 9.75 i.e., the families with 4 children in it, where as the least percentage i.e., 0 and 7.5 constitutes to 5 and 0 children respectively. 

This clearly states that number of children in a family clearly has an effect on the on-time repayment of loan. However it cannot be classified into any category of less children means earlier repayment or not.

- Is there a relation between marital status and repaying a loan on time?

In [44]:
family_status_with_debt = family_status_pivot['debt']*100
family_status_with_debt 

family_status
civil partnership    9.288963
divorced             7.112971
married              7.520194
unmarried            9.740491
widow / widower      6.562500
Name: debt, dtype: float64

### Conclusion

As per the above results, it's pretty clear that the families with 'family_status' as 'unmarried' have highest percentage of debts to be cleared i.e., 9.74% where as the lowest percentage being 6.5 that are the 'widow/widower' category. It can be seen that the relationship status in family definitely has some impact on the repayment of debts promptly.

- Is there a relation between income level and repaying a loan on time?

In [45]:
income_with_debt = income_pivot['debt']*100
income_with_debt

total_income_category
high income    8.055040
less income    8.138316
Name: debt, dtype: float64

### Conclusion

As per the above results, the precentages of families with debts in both the high income and low income categories are almost equal. In my view and according to the provided data, the amount of total income in a family doesn't really have any impact on repaying a loan on time.

- How do different loan purposes affect on-time repayment of the loan?

In [47]:
bank_data = bank_data.apply(getStem,axis=1) #applying the function to dataframe 


In [48]:
bank_data['purpose_stem'].value_counts()

construct    10840
car           4315
educ          3526
wed           2348
Name: purpose_stem, dtype: int64

In [50]:
purpose_pivot = bank_data.pivot_table(index='purpose_stem',values='debt',aggfunc='mean')
purpose_pivot

debt
purpose_stem          
car           0.093395
construct     0.072140
educ          0.092740
wed           0.079216

In [52]:
purpose_with_debt = purpose_pivot['debt']*100
purpose_with_debt

purpose_stem
car          9.339513
construct    7.214022
educ         9.273965
wed          7.921635
Name: debt, dtype: float64

### Conclusion

From the above output, it can be said that the purpose of purchase of car or something related to car has the highest percentage of debts to be paid i.e., 9.33% where as the lowest percentage being 7.21 contributing to the construction/real estate/house/housing purpose. The purpose of purchase has proven to be impactful over the customer's prompt repayment of loan.

### Step 4. General conclusion

Overall, there are several factors which when combined collectively can hugely impact the repayment of a loan on time. However as per the data provided by the bank it is evident that:

1. The on-time repayment of loan is independent of the  amount of income earned by the family.

2. And is dependent on the number of children, family status and purpose of purchase.

However, the impact of these factors might not be huge in number.